### PR TITLE
[RFC] Polyfill js for legacy browsers (adds events and classlist polyfills)

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
### Summary of Changes

As discussed in https://github.com/joomla/joomla-cms/pull/11585 this adds the ability to polyfill js events in older browsers (ex: IE8).
### Testing Instructions
- Apply https://github.com/joomla/joomla-cms/pull/11289 and them this patch
- Add this vanilla js events code in isis index.php

``` php
JHtml::_('behavior.polyfill', 'event', 'lt IE 9');
JHtml::_('behavior.polyfill', 'classlist', 'lt IE 11');

$this->addScriptDeclaration("
document.addEventListener('DOMContentLoaded', function() {

    this.getElementById('status').addEventListener('click', function() {
        alert('clicked the status bar');
    });

    this.querySelector('.container-logo').addEventListener('click', function() {
        alert('clicked the logo');
    });

    alert(this.body.classList);
});
");
```
- Refresh an admin page you will get an alert message with the body element classes
- Also, as the code says click the Joomla login in the admin and the admin status bar and check if you have a js alert.
- Test in all Joomla 3.x supported browsers. https://docs.joomla.org/Joomla_Browser_Support
- Code review
### Documentation Changes Required

Don't know. Usage examples:

``` php
JHtml::_('behavior.polyfill', 'event');
JHtml::_('behavior.polyfill', 'event', 'lt IE 9');
JHtml::_('behavior.polyfill', 'classlist', 'lt IE 11');
JHtml::_('behavior.polyfill', array('classlist'), 'lt IE 11');
JHtml::_('behavior.polyfill', array('event', 'classlist'));
JHtml::_('behavior.polyfill', array('event', 'classlist'), 'lte IE 11');
```
### Notes

Mantainers, this PR depends on https://github.com/joomla/joomla-cms/pull/11585

@Fedik @dgt41 is this it?

### CODE REVIEW

File: libraries/cms/html/behavior.php [hunk 1]
```diff
@@ -960,4 +960,51 @@ public static function tabstate()
...
+				continue;
+			}
+
+			// If include according to browser.
+			if ($conditionalBrowser)
+			{
+				// Get script path.
+				$doc        = JFactory::getDocument();
+				$scriptPath = JHtml::_('script', 'system/polyfill.' . $polyfillType . '.js', false, true, true, false, true);
+				$scriptType = !$doc->isHtml5() ? ' type="text/javascript"' : '';
```
Comment (Fedik) [inline libraries/cms/html/behavior.php]: IE do not understand you 😄

**Comment by Fedik** [comment]
yes, looks like 😉

File: libraries/cms/html/behavior.php (hunk 1 above)
Comment (andrepereiradasilva) [inline libraries/cms/html/behavior.php]: IE8 in emulation mode understood :smile:

**Comment by dgt41** [comment]
Can we add also classList, so I can drop all the polyfill stuff from the new calendar?

**Comment by Fedik** [comment]
and maybe `dataset` ?

**Comment by andrepereiradasilva** [comment]
Make PR for my `andrepereiradasilva:polyfill-behaviour` branch

**Comment by dgt41** [comment]
One more thing: how about some logic (check if `[if ]` exists) to move the tag before the other scripts (on head.php)?

**Comment by andrepereiradasilva** [comment]
JDocument::addScript and JDocument::addScriptVersion, and JHtml::script need to support conditional statements somehow.

**Comment by dgt41** [comment]
dataset and classList: https://github.com/remy/polyfills
@andrepereiradasilva @Fedik do you agree?

**Comment by Fedik** [comment]
"conditional statements" it is old stuff that should burn in fire 😄

**Comment by andrepereiradasilva** [comment]
@dgt41 the Ft was a nice API to build the js 

Check
- https://cdn.polyfill.io/v2/docs/examples
- https://cdn.polyfill.io/v2/docs/features/
- https://cdn.polyfill.io/v2/docs/api

If that library supports dataset and classList i think we should stick to it.

**Comment by dgt41** [comment]
I really don;t mind where the code comes from as long as it works and it well written.
So I was checking FT:
Element.prototype.classList is also not supported in IE9. Great More IE fun...

**Comment by andrepereiradasilva** [comment]
> I really don;t mind where the code comes from as long as it works and it well written.

i thnk the same but for me is better mantainence if we could use the same api to have the polyfills.

**Comment by dgt41** [comment]
dataset is not provided by FT, but for me this is not essential as you can still do 
```js
document.querySelector('.something').getAttribute('data-fantastic');
```
Which isn't that bad even compared to jQuery
```js
jQuery('.something').data('fantastic');
```

**Comment by andrepereiradasilva** [comment]
> Element.prototype.classList is also not supported in IE9. Great More IE fun...

The FT library has polyfill for that.

image

10, 11 seems to have partial support

**Comment by andrepereiradasilva** [comment]
classList polyfill added.

File: libraries/cms/html/behavior.php [hunk 2]
```diff
@@ -960,4 +960,51 @@ public static function tabstate()
...
+	 *
+	 * @param   string|array  $polyfillTypes       The polyfill type(s). Defaults to event.
+	 * @param   array         $conditionalBrowser  A IE conditional expression. Defaults to lt IE 9 (lower than IE 9).
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function polyfill($polyfillTypes = 'event', $conditionalBrowser = 'lt IE 9')
+	{
```
Comment (andrepereiradasilva) [inline libraries/cms/html/behavior.php]: why?

**Comment by dgt41** [comment]
@andrepereiradasilva can I propose changing the minified files to use .min.js and the uncompressed palin .js.
This -uncompressed.js really hurts my eyes 😛

**Comment by andrepereiradasilva** [comment]
> can I propose changing the minified files to use .min.js and the uncompressed palin .js.
> This -uncompressed.js really hurts my eyes :stuck_out_tongue:

lol i followed the same logic of all files in system/js dir
See https://github.com/joomla/joomla-cms/tree/staging/media/system/js
IMO if you want to change that you need to make a PR for change them all or you can use sunglasses :wink: .

File: libraries/cms/html/behavior.php (hunk 2 above)
Comment (dgt41) [inline libraries/cms/html/behavior.php]: classList partial implementation in IE 10-9

Comment (andrepereiradasilva) [inline libraries/cms/html/behavior.php]: i actually think it's better to leave the parameters empty by default (add if you want when calling)
i will remove the default values.

**Comment by dgt41** [comment]
It can't be done in 3.x otherwise I would have done that PR, but for J4 I will and also separate the joomla (system) from the external ones

**Comment by andrepereiradasilva** [comment]
ok so let's levae that for 4.x

**Comment by andrepereiradasilva** [comment]
default values removed:
```php
JHtml::_('behavior.polyfill', 'classlist');
JHtml::_('behavior.polyfill', 'classlist', 'lt IE 11');
JHtml::_('behavior.polyfill', array('classlist'), 'lt IE 11');
JHtml::_('behavior.polyfill', array('event', 'classlist'));
JHtml::_('behavior.polyfill', array('event', 'classlist'), 'lt IE 11');
```

**Comment by Fedik** [comment]
nativ `dataset` works like:
```html
<div data-foo-bar="blabla" class="something"></div>
```
```javascript
var element  = document.querySelector('.something')[0];
console.log(element.dataset.fooBar);
//or
console.log(element.dataset['foo-bar']);
```

**Comment by andrepereiradasilva** [comment]
http://caniuse.com/#feat=dataset

**Comment by Fedik** [comment]
was just note, for now can stay with `getAttribute`

**Comment by andrepereiradasilva** [comment]
so for this to work properly, we reallly need JDocument::addScript and JDocument::addScriptVersion, and JHtml::script to accept conditional statements
hum... i guess i have to review my https://github.com/joomla/joomla-cms/pull/11289 PR yet again ...

**Comment by andrepereiradasilva** [comment]
ok so i reviewed my PR again to accept conditinal statements.
https://github.com/joomla/joomla-cms/pull/11289

Please test careful since that PR is a sigificant refactor of methods to adding scripts to document.

**Comment by andrepereiradasilva** [comment]
updated test instructions, this PR now depends on https://github.com/joomla/joomla-cms/pull/11289

**Comment by andrepereiradasilva** [comment]
@Fedik @dgt41 need your help here.

Somehow the polyfills are not working in IE8 because of the first/last lines added by the polyfill.io generator https://github.com/joomla/joomla-cms/pull/11686/files#diff-0d0a277d1b9b770848b694a026f09fd4R480
```js
(function(undefined) {
// rest of code
})
.call('object' === typeof window && window || 'object' === typeof self && self || 'object' === typeof global && global || {});
```

any toughts?

example: https://cdn.polyfill.io/v2/polyfill.js?features=Element.prototype.classList&flags=always,gated

**Comment by andrepereiradasilva** [comment]
ok, i was wrong. so i discovered where the issue cames from.

the problem is the https://github.com/joomla/joomla-cms/pull/11686/files#diff-0d0a277d1b9b770848b694a026f09fd4R465 check

```if (!('addEventListener' in this)) {```

does not makes sens because we already added the addEventListener a few lines above

**Comment by Fedik** [comment]
sorry I was busy with other thing 😄 
seem you need to check for `addEventListener` somewhere earlier

**Comment by andrepereiradasilva** [comment]
yeah i know byt don'r really want to change the code form the polyfill service, if i change it's harder to update later.
This file is from https://cdn.polyfill.io/v2/polyfill.js?features=Event.DOMContentLoaded&flags=always,gated

**Comment by Fedik** [comment]
I think you can check for opposite thing
```javascript
if (this.attachEvent) {
  // it is IE
}
```

**Comment by andrepereiradasilva** [comment]
hum i think attachEvent is also availabe in IE9 and IE10.

**Comment by andrepereiradasilva** [comment]
ok it seems `if (!('createEvent' in document)) {` works
http://caniuse.com/#feat=customevent

**Comment by andrepereiradasilva** [comment]
seems to work fine.

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on c8781baa2b88b2237ee33d0e186b56471f552bfb<br /><br />I have tested it in IE8/9 (emulation mode)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11686">issues.joomla.org/joomla-cms/11686</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@dgt41 can you test this so it gets RTC?

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on c8781baa2b88b2237ee33d0e186b56471f552bfb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11686">issues.joomla.org/joomla-cms/11686</a>.</sub>

**Comment by dgt41** [comment]
screen shot 2016-08-31 at 17 56 48

RTC

**Comment by dgt41** [comment]
NOTE: https://github.com/joomla/joomla-cms/pull/11289 needs to be merged before this one

**Comment by wilsonge** [comment]
Merged with https://github.com/joomla/joomla-cms/commit/075a14b1ab4f89339fc03a9130a61bffcd0082a3

**Comment by zero-24** [comment]
@andrepereiradasilva can this work without: https://github.com/joomla/joomla-cms/pull/11289 ?

**Comment by andrepereiradasilva** [comment]
no it can't. https://github.com/joomla/joomla-cms/pull/11289 adds the hability to add IE contitional statemens tags and polyfill needs it.

**Pull request closed without merging**